### ISSUE TITLE
Wish for API.run((r) -> {})  method

### ISSUE THREAD

**Issue by iliax** [vavr-io/vavr#1218]
Hi, thanks a lot for your library! I trying hard to use it in my current projects, and it's really rocks) But I found one limitation(or am I doing smth wrong) - using math this way I have type cast in Case method parameter (castedToIng, castToDouble) 

```
String res = Match(obj).of(
                Case(instanceOf(Integer.class), castedToInt -> { /* do smth with Int val */ }),
                Case(instanceOf(Double.class), castToDouble -> { /* do smth with Double val */ }}),
                Case($(), o -> { throw new NumberFormatException(); })
        );
```

But using run(), with no parametrized lambda I cant use type cast. In other words, why is it impossible to do it like 

```
Match(obj).of(
                Case(instanceOf(Integer.class), run((castedToInt) -> { /* do smth with Int val */  })),
                Case(instanceOf(Double.class), run((castToDouble) -> { /* do smth with Double val */  })),
                Case($(), o -> { throw new NumberFormatException(); })
        );
```

**Comment by danieldietrich** [comment]
Hi @iliax, thx for using Javaslang!

In your example, the `Case(Pattern1, R)` methods are called (first two cases). Our run() currently takes only a Runnable. Given that, it is possible like this:

```java
    @Test
    public void shouldRunWithInferredArguments() {

        class OuterWorld {

            Number effect = null;

            void writeInt(int i) {
                effect = i;
            }

            void writeDouble(double d) {
                effect = d;
            }

        }

        final OuterWorld outerWorld = new OuterWorld();
        final Object obj = .1d;

        Match(obj).of(
                Case(instanceOf(Integer.class), i -> run(() -> outerWorld.writeInt(i))),
                Case(instanceOf(Double.class), d -> run(() -> outerWorld.writeDouble(d))),
                Case($(), run(() -> { throw new NumberFormatException(); }))
        ).run(); // <-- only works with current master because of a bug, see below

        assertThat(outerWorld.effect).isEqualTo(.1d);
    }
```

I like your suggestion. It involves adding new run-methods

* run(Function)
* run(BiFunction)
* run(...)

But I think these methods will be ambiguous. The compiler can't figure out which method to take when we provide a lambda or method reference. I will test a bit with it but I think it is not possible. That means that my first example is the way to go.

**NOTE: Please note that there is currently a bug that run() eagerly evaluates. It will be solved with Javaslang 2.0.2 today. See #1216**

**Comment by danieldietrich** [comment]
It seems to be possible what you suggested:

```java
    static <T> Function<T, Void> run(Consumer<? super T> consumer) {
        return t -> {
            consumer.accept(t);
            return null;
        };
    }

    static <T1, T2> BiFunction<T1, T2, Void> run(BiConsumer<? super T1, ? super T2> consumer) {
        return (t1, t2) -> {
            consumer.accept(t1, t2);
            return null;
        };
    }

    // more run methods (Consumer3, ... Consumer8)

    void test() {
        Match(new Object()).of(
                Case(instanceOf(Integer.class), run(i -> System.out.println(i))),
                Case(instanceOf(Double.class), run(d -> System.out.println(d))),
                Case($(), run(ignored -> { throw new NumberFormatException(); }))
        );
    }
```

Now we have the option to drop the existing run method. and add these. I will wait providing the fix for #1216 . This API change has to be well thought through.

**Comment by danieldietrich** [comment]
Match is an expression - that's the idea of functional programming.
It is not our intention to push side-effecting / imperative programming. That's the wrong direction and will lead to bad code.
